jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java

```diff
@@ -392,83 +404,153 @@ public abstract File getFile()
+        {
+            List<String> entries = new ArrayList<>();
+            for (Path entry : dir)
+            {
+                String name = entry.getFileName().toString();
 
-    // FIXME: this appears to not be used
-    @SuppressWarnings("javadoc")
-    public Object getAssociate()
-    {
-        return _associate;
+                if (Files.isDirectory(entry))
+                {
+                    name += "/";
+                }
+
+                entries.add(name);
+            }
+            return entries;
+        }
+        catch (DirectoryIteratorException e)
+        {
+            LOG.debug("Directory list failure", e);
+        }
+        catch (IOException e)
+        {
+            LOG.debug("Directory list access failure", e);
+        }
+        return null;
     }
 
-    // FIXME: this appear to not be used
-    @SuppressWarnings("javadoc")
-    public void setAssociate(Object o)
+    public Resource resolve(String subPath) throws IOException
     {
-        _associate = o;
+        // Check that the path is within the root,
+        // but use the original path to create the
+        // resource, to preserve aliasing.
+        if (URIUtil.canonicalPath(subPath) == null)
+            throw new MalformedURLException(subPath);
+
+        if (URIUtil.SLASH.equals(subPath))
+            return this;
+
+        // Sub-paths are always resolved under the given URI,
+        // we compensate for input sub-paths like "/subdir"
+        // where default resolve behavior would be to treat
+        // that like an absolute path.
+        while (subPath.startsWith(URIUtil.SLASH))
+            subPath = subPath.substring(1);
+
+        URI uri = getURI();
+        URI resolvedUri;
+        if (uri.isOpaque())
+        {
+            // TODO create a subclass with an optimized implementation of this.
+            // The 'jar:file:/some/path/my.jar!/foo/bar' URI is opaque b/c the jar: scheme is not followed by //
+            // so we take the scheme-specific part (i.e.: file:/some/path/my.jar!/foo/bar) and interpret it as a URI,
+            // use it to resolve the subPath then re-prepend the jar: scheme before re-creating the URI.
+            String scheme = uri.getScheme();
+            URI subUri = URI.create(uri.getSchemeSpecificPart());
+            if (subUri.isOpaque())
+                throw new IllegalArgumentException("Unsupported doubly opaque URI: " + uri);
+
+            if (!subUri.getPath().endsWith(URIUtil.SLASH))
+                subUri = URI.create(subUri + URIUtil.SLASH);
+
+            URI subUriResolved = subUri.resolve(subPath);
+            resolvedUri = URI.create(scheme + ":" + subUriResolved);
+        }
+        else
+        {
+            if (!uri.getPath().endsWith(URIUtil.SLASH))
+                uri = URI.create(uri + URIUtil.SLASH);
+            resolvedUri = uri.resolve(subPath);
+        }
+        return newResource(resolvedUri);
```

Comment:
Isn't the following simpler and more efficient?
Also the toString could be a final field, to make it even better
```suggestion
        return newResource(URI.create(URIUtil.addPaths(getURI().toString(), subPath)));
```

Reply:
Strangely enough, in my experiments, I am seeing a different between:
```java
        Path.of(URI.create("file:/tmp/swedish-ä.txt"));
        Path.of(URI.create("file:///tmp/swedish-ä.txt"));
```

The second of these throws:
```
java.lang.IllegalArgumentException: Bad escape
	at java.base/sun.nio.fs.UnixUriUtils.fromUri(UnixUriUtils.java:88)
	at java.base/sun.nio.fs.UnixFileSystemProvider.getPath(UnixFileSystemProvider.java:102)
	at java.base/java.nio.file.Path.of(Path.java:203)
```

So it looks like we will need to normalize how we handle "file:///"

Also strangely enough the code:
```java
        System.err.println(URI.create("file:///tmp/").resolve("swedish-ä.txt"));
```
produces: `file:/tmp/swedish-ä.txt`

Reply:
Doesn't using `URIUtil.addPaths` defeat the purpose of this `resolve()` change as we're not relying on `URI.resolve()` anymore to check for the validity of the URI?

Reply:
Well, we still end up calling `URI.create()` onto the string created by `URIUtil` so maybe that's still alright.

But there are other changes as well: `Resource.newResource("file:/a/b/c").resolve("//foo")` results in `file:/a/b/c//foo` instead of `file:/a/b/c/foo`.

Reply:
It's still resolved in URI space and is checked by URI. I think either way we want to think about those differences.

Reply:
> ```java
>         Path.of(URI.create("file:/tmp/swedish-ä.txt"));
>         Path.of(URI.create("file:///tmp/swedish-ä.txt"));
> ```

The first one is the correct one per URI spec.
The second one is for `java.io.File` compatibility reasons. (Don't use this one, as it just keeps all of the bad bugs in `File` alive in Jetty 12)

Internally, the `sun.nio.fs.UnixUriUtils` will use `new File(uri).toPath()` if the uri starts with `file:///`.

We should strip back any `file:///` to just `file:/` in Jetty 12.

Reply:
+1

Reply:
Sorry, I had a typo in my message, I edited to show the behavior correctly.